<div style="border:solid black 2px; padding: 20px"><h1 style="color:black; margin-bottom:20px">Reviewer's comment v1</h1>
Halo Ahmad !

Perkenalkan saya Code Reviewers, disini saya akan mereview tugas Ahmad ya.

Saya akan memberikan beberapa komen seperti dibawah ini, tolong jangan dipindah, dirubah, maupun dihapus ya.

Komen yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-warning">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Terdapat beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan

</div>

Dan tentu saja Ahmad dapat menjawab saya dengan menggunakan ini

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak
</div>

# CREDIT ASSESSMENT

# Menganalisis risiko gagal bayar peminjam

Tugasmu adalah menyiapkan laporan untuk divisi kredit suatu bank. Kamu akan mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak yang dimilikinya terhadap probabilitas gagal bayar dalam pelunasan pinjaman. Pihak bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.

Laporanmu akan dipertimbangkan pada saat membuat **penilaian kredit** untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

[Dalam notebook ini, kamu dapat menemukan petunjuk dan instruksi singkat serta arahan untuk berpikir. Jangan abaikan petunjuk tersebut karena mereka dirancang untuk membekalimu dengan struktur pengerjaan proyek dan akan membantumu melakukan analisis secara lebih mendalam. Sebelum menyerahkan proyekmu, pastikan kamu menghapus semua petunjuk dan deskripsi yang diberikan kepadamu. Sebaliknya, buat laporan ini terlihat seolah-olah dikirimkan oleh kamu kepada rekan kerja untuk menunjukkan temuanmu - mereka tidak perlu tahu bahwa kamu mendapatkan bantuan eksternal dari kami! Untuk membantumu, kami menempatkan semua petunjuk yang harus kamu hapus dalam tanda kurung siku.]

[Sebelum kamu masuk ke dalam proses analisis datamu, jelaskan tujuan proyek dan hipotesis yang akan kamu uji.]

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Buka file data dan baca informasi umumnya.

[Mulailah dengan mengimpor library dan memuat datanya. Kamu mungkin akan menyadari bahwa kamu memerlukan library tambahan saat kamu mengerjakan proyek ini dan itu merupakan hal yang normal. Hanya saja, pastikan untuk memperbarui bagian ini jika dibutuhkan.]

In [1]:
# Muat semua library
import pandas as pd

# Muat datanya
df = pd.read_csv('/datasets/credit_scoring_eng.csv')

## Soal 1. Eksplorasi data

**Deskripsi data**
- `children` - jumlah anak dalam keluarga
- `days_employed` - pengalaman kerja nasabah dalam hari
- `dob_years` - usia nasabah dalam tahun
- `education` - tingkat pendidikan nasabah
- `education_id` - pengidentifikasi untuk tingkat pendidikan nasabah
- `family_status` - status perkawinan
- `family_status_id` - pengidentifikasi untuk status perkawinan nasabah
- `gender` - jenis kelamin nasabah
- `income_type` - jenis pekerjaan
- `debt` - apakah nasabah pernah melakukan gagal bayar pinjaman
- `total_income` - pendapatan bulanan
- `purpose` - tujuan mendapatkan pinjaman

[Sekarang, saatnya menjelajahi data kita. Kamu perlu melihat berapa banyak kolom dan baris yang dimiliki oleh data, serta mencermati beberapa baris data untuk memeriksa potensi masalah dengan data.]

In [2]:
# Mari kita lihat berapa banyak baris dan kolom yang dimiliki oleh dataset kita
df.shape

(21525, 12)

In [3]:
# Mari tampilkan N baris pertama
df.head(10)


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

 [Jelaskan apa yang kamu lihat dan perhatikan dalam sampel data yang kamu tampilkan. Apakah terdapat masalah yang mungkin memerlukan penyelidikan dan perubahan lebih lanjut?]
1. Di kolom 'days_employed' terdapat nilai yang negatif, padahal kolom 'days_employed' adalah jumlah hari kerja dan tidak mungkin nilainya negatif.
2. Di kolom 'days_employed' tersapat nilai yang terlalu besar dan tidak wajar yaitu 340266.072047
3. Di kolom 'education', penulisan datanya belum seragam
4. Di kolom 'purpose' masih mengandung duplikat implisit

In [4]:
# Dapatkan informasi data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


[Apakah terdapat nilai yang hilang di semua kolom atau hanya di beberapa kolom? Jelaskan secara singkat apa yang kamu lihat dalam 1-2 kalimat.]

#Terdapat nilai yang hilang di dua kolom yaitu kolom days_employed dan total_income

In [5]:
# Mari kita lihat tabel yang telah difilter dengan nilai yang hilang di kolom pertama yang mengandung data yang hilang
missing_df = df.loc[df['days_employed'].isnull()]
missing_df.isnull().sum()


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

[Apakah nilai yang hilang tampak simetris? Apakah kita yakin dengan asumsi ini? Jelaskan pendapatmu secara singkat di bagian ini. Kamu mungkin ingin melakukan penyelidikan lebih lanjut dan menghitung nilai yang hilang di semua baris dengan nilai yang hilang untuk memastikan bahwa sampel yang hilang memiliki ukuran yang sama.]

In [6]:
df.isnull().sum()
## missing value nya simetris karena setelah difilter di kolom days empoloyed, jumlah missing value di kolom total income juga sama

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [7]:
# Mari kita terapkan beberapa kondisi untuk memfilter data dan melihat jumlah baris dalam tabel yang telah difilter.
missing_df = df.loc[(df['days_employed'].isnull()) & (df['total_income'].isnull())]
missing_df


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Kesimpulan sementara**

[Apakah jumlah baris dalam tabel yang telah difilter sesuai dengan jumlah nilai yang hilang? Kesimpulan apa yang bisa kita buat dari hal ini?]

[Hitung persentase nilai yang hilang jika dibandingkan dengan keseluruhan dataset. Apakah nilai yang hilang merupakan bagian data yang cukup besar? Jika demikian, kamu sebaiknya perlu mengisi nilai yang hilang. Untuk melakukannya, pertama-tama kita harus mempertimbangkan apakah data yang hilang bisa jadi disebabkan oleh karakteristik nasabah tertentu, seperti jenis pekerjaan atau yang lainnya. Kamu harus memutuskan karakteristik mana yang menurut *kamu* mungkin merupakan penyebabnya. Kedua, kita harus memeriksa apakah ada ketergantungan nilai yang hilang pada nilai indikator lain dengan kolom-kolom yang mengidentifikasikan karakteristik tertentu nasabah.]

[Jelaskan langkahmu selanjutnya dan bagaimana korelasinya dengan kesimpulan yang kamu buat sejauh ini.]

In [8]:
# Mari kita periksa nasabah yang tidak memiliki data tentang karakteristik yang teridentifikasi dan kolom dengan nilai yang hilang
missing_df.isnull().sum() / df.shape[0] * 100

#persentase nilai yang hilang sebesar 10.099884 %


children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

In [11]:
# Periksalah distribusinya

# cek distribusi di kolom education_id
missing_df['education_id'].value_counts(normalize=True)

1    0.708372
0    0.250230
2    0.031739
3    0.009660
Name: education_id, dtype: float64

In [15]:
df['education_id'].value_counts(normalize=True)

1    0.707689
0    0.244367
2    0.034564
3    0.013101
4    0.000279
Name: education_id, dtype: float64

[Deksripsikan temuanmu di sini.]

**Kemungkinan penyebab hilangnya nilai dalam data**

[Ajukan gagasanmu mengenai kemungkinan alasan hilangnya nilai-nilai tersebut. Apakah menurutmu mereka hilang secara acak atau apakah terdapat pola-pola tertentu?]

[Mari kita mulai memeriksa apakah nilai yang hilang bersifat acak.]

In [16]:
# Memeriksa distribusi di seluruh dataset
# Memeriksa distribusi di kolom gender
missing_df['gender'].value_counts(normalize=True)


F    0.682613
M    0.317387
Name: gender, dtype: float64

In [17]:
df['gender'].value_counts(normalize=True)


F      0.661370
M      0.338583
XNA    0.000046
Name: gender, dtype: float64

In [18]:
# Memeriksa distribusi di kolom children
missing_df['children'].value_counts(normalize=True)

 0     0.661914
 1     0.218491
 2     0.093836
 3     0.016559
 20    0.004140
 4     0.003220
-1     0.001380
 5     0.000460
Name: children, dtype: float64

In [19]:
df['children'].value_counts(normalize=True)

 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

**Kesimpulan sementara**

[Apakah distribusi dalam dataset yang asli mirip dengan distribusi tabel yang telah difilter? Apa artinya hal tersebut untuk kita?]

[Jika menurutmu kita belum dapat membuat kesimpulan apa pun, mari kita kembali menyelidiki dataset lebih lanjut. Mari pikirkan alasan lain yang dapat menyebabkan data hilang dan periksa apakah kita dapat menemukan pola tertentu yang dapat membuat kita berpikir bahwa hilangnya nilai-nilai tersebut tidak terjadi secara acak. Karena ini merupakan tugasmu maka bagian ini adalah opsional.]

In [20]:
# Periksa penyebab dan pola lain yang dapat mengakibatkan nilai yang hilang
# Memeriksa distribusi di kolom family_status
missing_df['family_status'].value_counts(normalize=True)


married              0.568997
civil partnership    0.203312
unmarried            0.132475
divorced             0.051518
widow / widower      0.043698
Name: family_status, dtype: float64

In [21]:
df['family_status'].value_counts(normalize=True)

married              0.575145
civil partnership    0.194053
unmarried            0.130685
divorced             0.055517
widow / widower      0.044599
Name: family_status, dtype: float64

**Kesimpulan sementara**

[Apakah pada akhirnya kita dapat memastikan bahwa nilai yang hilang adalah suatu kebetulan? Periksa hal lain yang menurutmu penting di sini.]

In [22]:
# Periksa pola lainnya - jelaskan pola tersebut
# Memeriksa distribusi di kolom income_type
missing_df['income_type'].value_counts(normalize=True)

employee         0.508280
business         0.233671
retiree          0.189972
civil servant    0.067617
entrepreneur     0.000460
Name: income_type, dtype: float64

In [23]:
df['income_type'].value_counts(normalize=True)

employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
entrepreneur                   0.000093
unemployed                     0.000093
paternity / maternity leave    0.000046
student                        0.000046
Name: income_type, dtype: float64

**Kesimpulan**

[Apakah kamu menemukan suatu pola? Bagaimana kamu mendapatkan kesimpulan ini?]

[Jelaskan bagaimana kamu akan mengatasi nilai-nilai yang hilang. Pertimbangkan kategori dengan nilai yang hilang.]

[Rencanakan secara singkat langkahmu selanjutnya untuk mentransformasi data. Kamu mungkin perlu mengatasi berbagai jenis masalah: duplikat, pencatatan yang berbeda, sumber data yang salah, dan nilai yang hilang.]

-Apakah kamu menemukan suatu pola? Bagaimana kamu mendapatkan kesimpulan ini?

Distribusi di setiap kolom antara dataset dengan data yang telah difilter mirip karena nilainya hampir sama.

-Jelaskan bagaimana kamu akan mengatasi nilai-nilai yang hilang. Pertimbangkan kategori dengan nilai yang hilang.

Nilai yang hilang akan diganti dengan nilai rata-rata (mean) atau nilai tengah (median)

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Transformasi data

[Mari kita perhatikan setiap kolom untuk melihat masalah apa yang mungkin dimiliki mereka.]

[Mulailah dengan menghapus duplikat dan memperbaiki data tentang informasi pendidikan jika diperlukan.]

In [24]:
# Mari kita lihat semua nilai di kolom pendidikan untuk memeriksa ejaan apa yang perlu diperbaiki
df['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

In [25]:
# Perbaiki pencatatan jika diperlukan
df['education'] = df['education'].str.lower()

In [26]:
# Periksa semua nilai di kolom untuk memastikan bahwa kita telah memperbaikinya dengan tepat
df['education'].value_counts()


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

[Periksa data kolom `children`]

In [27]:
# Mari kita lihat distribusi nilai pada kolom `children`
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

[Apakah terdapat hal-hal yang aneh di kolom tersebut? Jika jawabannya iya, seberapa tinggi persentase data yang bermasalah? Bagaimana mereka bisa terjadi? Buat keputusan tentang apa yang akan kamu lakukan dengan data ini dan jelaskan alasannya.]

In [28]:
#Apakah terdapat hal-hal yang aneh di kolom tersebut?
# Ya, ada hal-hal yang aneh di kolom children yakni ada nilai 20 dan -1. Jumlah anak 20 sangat kecil kemungkinannya, dan nilai negatif tidak mungkin dalam kolom jumlah anak.

#Persentase data yang bermasalah
df['children'].value_counts()/ df.shape[0]*100

#Yang nilai 20 = 0.35% dan yang nilai -1 = 0.218%

#Bagaimana mereka bisa terjadi? Kemungkinan salah input data (human error)

#Buat keputusan tentang apa yang akan kamu lakukan dengan data ini dan jelaskan alasannya.
#Saya memutuskan untuk mengganti 20 dengan 2 dan -1 dengan 1 karena asumsi saya ini hanya human error


 0     65.732869
 1     22.383275
 2      9.547038
 3      1.533101
 20     0.353078
-1      0.218351
 4      0.190476
 5      0.041812
Name: children, dtype: float64

In [29]:
# [perbaiki data berdasarkan keputusanmu]
df['children'] = df['children'].replace(20, 2).replace(-1, 1)

In [30]:
# Periksa kembali kolom `children` untuk memastikan bahwa semuanya telah diperbaiki
df['children'].value_counts()


0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Periksa data dalam kolom `days_employed`. Pertama-tama, pikirkan tentang masalah apa yang mungkin ada pada kolom tersebut, serta pikirkan juga apa yang mungkin ingin kamu periksa dan bagaimana kamu akan melakukannya.]

In [31]:
# Temukan data yang bermasalah di kolom `days_employed` jika memang terdapat masalah dan hitung persentasenya
df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

[Jika jumlah data yang bermasalah tinggi, hal tersebut mungkin disebabkan oleh beberapa masalah teknis. Kita mungkin perlu mengusulkan alasan paling jelas mengapa hal tersebut dapat terjadi dan bagaimana seharusnya data yang benar, mengingat kita tidak dapat menghapus baris yang bermasalah ini.]

In [37]:
# Atasi nilai yang bermasalah, jika ada
# Ini Kemungkinan datanya masih dalam satuan jam, maka perlu dibagi 24
days_employ = df.loc[df['days_employed'] > 0, 'days_employed'] / 24
df.loc[df['days_employed'] > 0, 'days_employed'] = days_employ
days_employ.describe()

count    19351.000000
mean         0.148071
std          0.158436
min          0.001746
25%          0.045367
50%          0.087049
75%          0.198743
max          1.330219
Name: days_employed, dtype: float64

In [38]:
# Mengubah data negatif jadi positif
df['days_employed'] = abs(df['days_employed'])

In [39]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['days_employed'].describe()

count    19351.000000
mean         0.148071
std          0.158436
min          0.001746
25%          0.045367
50%          0.087049
75%          0.198743
max          1.330219
Name: days_employed, dtype: float64

In [40]:
df['days_employed'] / 365

0        0.001672
1        0.000798
2        0.001114
3        0.000817
4        0.000117
           ...   
21520    0.000898
21521    0.000118
21522    0.000419
21523    0.000617
21524    0.000393
Name: days_employed, Length: 21525, dtype: float64

<div class="alert alert-success">
<b>Code Reviewers's comment v.2</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

"Data di kolom 'days_employed' dibagi dengan 24 karena diasumsikan datanya masih dalam satuan jam, sehingga untuk mendapatkan data hari, harus dibagi dengan 24. Contohnya nilai max 401755.400475 (satuan jam), jika dibagi dengan 24 hasilnya adalah 16739.8084 (satuan hari). Jika 16739.8084 dibagi dengan 365 maka hasilnya adalah 45.862 tahun. Masih masuk akal ada orang yang bekerja selama 45 tahun."

[Sekarang, mari kita lihat usia nasabah dan mengecek apakah terdapat masalah di sana. Sekali lagi, pikirkan tentang kemungkinan kejanggalan apa yang bisa kita temui dalam kolom ini, misalnya angka usia yang tidak masuk akal.]

In [41]:
# Periksa `dob_years` untuk nilai yang mencurigakan dan hitung persentasenya
df['dob_years'].sort_values().unique()


array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

[Putuskan apa yang akan kamu lakukan dengan nilai yang bermasalah dan jelaskan alasannya.]

In [42]:
# Atasi masalah pada kolom `dob_years`, jika terdapat masalah
# Ada nilai 0 dan ini mustahil ada usia 0 jadi harus diubah ke nilai median

int(df['dob_years'].median())


42

In [40]:
# Mengubah nilai 0 menjadi median
median_age = int(df['dob_years'].median())
df['dob_years'] = df['dob_years'].replace(0, median_age)

In [43]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['dob_years'].sort_values().unique()


array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

[Sekarang, saatnya memeriksa kolom `family_status`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu kamu atasi.]

In [44]:
# Mari kita lihat nilai untuk kolom ini
df['family_status'].value_counts()


married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [35]:
# Atasi nilai yang bermasalah di `family_status`, jika ada
# Tidak ada masalah di kolom 'family_status'



In [36]:
# Periksa hasilnya - pastikan nilainya telah diperbaiki


[Sekarang, saatnya memeriksa kolom `gender`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu kamu atasi]

In [45]:
# Mari kita liat nilai dalam kolom ini
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [38]:
# Atasi nilai-nilai yang bermasalah, jika ada
# Ada nilai yang aneh yaitu XNA, tapi karena hanya ada satu baris maka diabaikan saja.

In [39]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki



[Sekarang, saatnya memeriksa kolom `income_type`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu kamu atasi]

In [46]:
# Mari kita lihat nilai dalam kolom ini
df['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [41]:
# Atasi nilai yang bermasalah, jika ada
# Tidak ada masalah di kolom 'income_type'

In [42]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki



[Sekarang, saatnya melihat apakah ada duplikat di dalam data kita. Jika kita menemukannya, kamu harus memutuskan apa yang akan kamu lakukan dengan duplikat tersebut dan menjelaskan alasannya.]

In [45]:
# Periksa duplikat
df.duplicated().sum()


72

In [47]:
# Atasi duplikat, jika ada
df.drop_duplicates(inplace=True)

In [48]:
# Lakukan pemeriksaan terakhir untuk mengecek apakah kita memiliki duplikat
df.duplicated().sum()

0

In [49]:
# Periksa ukuran dataset yang sekarang kamu miliki setelah manipulasi pertama yang kamu lakukan
df.head(10)

children  days_employed  dob_years            education  education_id  \
0         1       0.610364         42    bachelor's degree             0   
1         1       0.291146         36  secondary education             1   
2         0       0.406787         33  secondary education             1   
3         3       0.298376         32  secondary education             1   
4         0       0.042733         53  secondary education             1   
5         0       0.066998         27    bachelor's degree             0   
6         0       0.208276         43    bachelor's degree             0   
7         0       0.011052         50  secondary education             1   
8         2       0.501292         35    bachelor's degree             0   
9         0       0.158330         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

[Jelaskan dataset barumu: jelaskan secara singkat apa saja perubahannya dan seberapa besar persentase perubahannya, jika ada.]


In [50]:
# Jelaskan dataset barumu: jelaskan secara singkat apa saja perubahannya dan seberapa besar persentase perubahannya, jika ada.
# Pada dataset yang baru ada total 21453 baris sedangkan pada dataset awal ada 21525 baris berarti persentase perubahannya =
# (selisih baris / total baris awal) * 100 = 0,334 %

df.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21454 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21454 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

# Bekerja dengan nilai yang hilang

[Untuk mempercepat pekerjaan dengan sejumlah data, kamu mungkin ingin menggunakan dictionary untuk beberapa nilai yang memiliki ID. Jelaskan mengapa dan dictionary manakah yang akan kamu gunakan.]

In [52]:
# Temukan dictionary
# Dictionary untuk kolom education 
df.loc[:, ['education_id', 'education']].drop_duplicates()

education_id            education
0                0    bachelor's degree
1                1  secondary education
13               2         some college
31               3    primary education
2963             4      graduate degree

In [53]:
# Dictionary untuk kolom family_status 
df.loc[:, ['family_status_id', 'family_status']].drop_duplicates()

family_status_id      family_status
0                  0            married
4                  1  civil partnership
18                 2    widow / widower
19                 3           divorced
24                 4          unmarried

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Memperbaiki nilai yang hilang di `total_income`

[Jelaskan secara singkat kolom dengan nilai yang hilang manakah yang perlu kamu tangani. Jelaskan bagaimana kamu akan memperbaikinya.]


[Mulai dengan mengatasi total nilai pendapatan yang hilang. Buat kategori usia untuk nasabah. Buat kolom baru yang memuat kategori usia. Strategi ini dapat membantu untuk menghitung total nilai pendapatan.]


In [54]:
df['dob_years'].describe()

count    21454.000000
mean        43.271231
std         12.570822
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [55]:
# Mari kita tulis sebuah fungsi untuk menghitung kategori usia
def age_categorizing(age):
    if age <= 20:
        category = '0-20'
    elif age > 20 and age <= 40:
        category = '20-40'
    elif age > 40 and age <= 60:
        category = '40-60'
    elif age > 60 and age <= 80:
        category = '60-80'
    else:
        category = '>80'

    return category
    

In [56]:
# Lakukan pengujian untuk melihat apakah fungsimu bekerja atau tidak
age_categorizing(age=55)

'40-60'

In [59]:
# Buatlah kolom baru berdasarkan fungsi
df['age_category'] = df['dob_years'].apply(age_categorizing)


In [62]:
# Periksa bagaimana nilai di dalam kolom baru
df.loc[:, ['dob_years', 'age_category']].sample(n=20)


dob_years age_category
18943         29        20-40
11198         32        20-40
1160          60        40-60
19332         66        60-80
7853          23        20-40
18837         38        20-40
1624          57        40-60
13640         42        40-60
20282         46        40-60
7737          45        40-60
7330          52        40-60
2618          48        40-60
5699          37        20-40
6849          44        40-60
19994         44        40-60
14370         64        60-80
2268          34        20-40
15007         36        20-40
20806         47        40-60
13378         26        20-40

[Pikirkan tentang faktor-faktor yang biasanya bergantung pada pendapatan. Pada akhirnya, kamu akan mengetahui apakah kamu harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Untuk membuat keputusan ini, kamu mungkin ingin melihat distribusi faktor-faktor yang kamu identifikasi sebagai dampak dari pendapatan seseorang.]

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Buatlah tabel yang hanya memuat data tanpa nilai yang hilang. Data ini akan digunakan untuk memperbaiki nilai yang hilang.]

In [63]:
# Buat tabel tanpa nilai yang hilang dan tampilkan beberapa barisnya untuk memastikan semuanya berjalan dengan baik
df.loc[~df['total_income'].isnull()]

children  days_employed  dob_years            education  education_id  \
0             1       0.610364         42    bachelor's degree             0   
1             1       0.291146         36  secondary education             1   
2             0       0.406787         33  secondary education             1   
3             3       0.298376         32  secondary education             1   
4             0       0.042733         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1       0.327642         43  secondary education             1   
21521         0       0.043194         67  secondary education             1   
21522         1       0.152875         38  secondary education             1   
21523         3       0.225151         38  secondary education             1   
21524         2       0.143555         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose age_category  
0         40620.102    purchase of the house        40-60  
1         17932.802             car purchase        20-40  
2         23341.752    purchase of the house        20-40  
3         42820.568  supplementary education        20-40  
4         25378.572        to have a wedding        40-60  
...             ...                      ...          ...  
21520     35966.698     housing transactions        40-60  
21521     24959.969        purchase of a car        60-80  
21522     14347.610                 property        20-40  
21523     39054.888        buying my own car        20-40  
21524     13127.587             to buy a car        20-40  

[19351 rows x 13 columns]

In [64]:
nomiss_data = df.loc[~df['total_income'].isnull()]

In [65]:
# Perhatikan nilai rata-rata untuk pendapatan berdasarkan faktor yang telah kamu identifikasi
nomiss_data.groupby('age_category')['total_income'].mean().reset_index()

age_category  total_income
0         0-20  23073.283620
1        20-40  27421.779199
2        40-60  27052.238111
3        60-80  23057.777452

In [62]:
# Perhatikan nilai median untuk pendapatan berdasarkan faktor yang telah kamu identifikasi
nomiss_data.groupby('age_category')['total_income'].median().reset_index()

age_category  total_income
0         0-20    17257.2770
1        20-40    23972.9590
2        40-60    23219.8075
3        60-80    19637.0560

[Ulangi perbandingan tersebut untuk beberapa faktor. Pastikan kamu mempertimbangkan berbagai aspek dan menjelaskan proses berpikirmu.]



[Buat keputusan tentang karakteristik yang paling menentukan pendapatan dan apakah kamu akan menggunakan median atau rata-rata. Jelaskan mengapa kamu membuat keputusan ini]


In [63]:
#  Tulis fungsi yang akan kita gunakan untuk mengisi nilai yang hilang
# Saya akan mengisi missing value berdasarkan median kolom total income di setiap age category
median = pd.pivot_table(df, columns='age_category', values='total_income', aggfunc='median')
median        
        

age_category       0-20      20-40       40-60      60-80
total_income  17257.277  23972.959  23219.8075  19637.056

In [64]:
def group_fillna(dataframe, group_col, missing_col):
    median = pd.pivot_table(df, columns=group_col, values=missing_col, aggfunc='median')

    for col in median.columns:
        df.loc[(df[group_col]==col) & (df[missing_col].isnull()), missing_col] = median[col][0]

    return df

In [65]:
# Memeriksa bagaimana nilai di dalam kolom baru
df = group_fillna(dataframe=df, group_col='age_category', missing_col='total_income')
df.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0      24.614155         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose age_category  
0    purchase of the house        40-60  
1             car purchase        20-40  
2    purchase of the house        20-40  
3  supplementary education        20-40  
4        to have a wedding        40-60

In [62]:
# Terapkan fungsi tersebut ke setiap baris


In [66]:
# Periksa apakah kita mendapatkan kesalahan
df.isnull().sum()

children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_category           0
dtype: int64

[Jika kamu menemukan kesalahan dalam menyiapkan nilai untuk data yang hilang, kemungkinan ada sesuatu yang istimewa tentang data untuk kategori tersebut. Coba pikirkan - kamu mungkin ingin memperbaiki beberapa hal secara manual, jika terdapat cukup data untuk menemukan median/rata-rata.]


In [64]:
# Ganti nilai yang hilang jika terdapat kesalahan


[Setelah kamu selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

In [67]:
# Periksa jumlah entri di kolom
df.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21453 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21453 non-null  int64  
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  float64
 11  purpose           21453 non-null  object 
 12  age_category      21453 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.3+ MB


<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

###  Memperbaiki nilai di `days_employed`

[Pikirkan tentang parameter yang dapat membantumu memperbaiki nilai yang hilang di kolom ini. Pada akhirnya, kamu akan mengetahui apakah kamu harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Kamu mungkin perlu melakukan penelitian yang sama dengan yang kamu lakukan saat memperbaiki data di kolom sebelumnya.]

In [68]:
# Distribusi median dari `days_employed` berdasarkan parameter yang kamu identifikasi
median = pd.pivot_table(df, columns='income_type', values='days_employed', aggfunc='median')
median



income_type       business  civil servant     employee  entrepreneur  \
days_employed  1547.382223    2689.368353  1574.202821    520.848083   

income_type    paternity / maternity leave    retiree     student  unemployed  
days_employed                  3296.759962  26.418787  578.751554   26.505617

In [69]:
# Distribusi rata-rata dari `days_employed` berdasarkan parameter yang kamu identifikasi
mean = pd.pivot_table(df, columns='income_type', values='days_employed', aggfunc='mean')
mean

income_type       business  civil servant     employee  entrepreneur  \
days_employed  2111.524398    3399.896902  2326.499216    520.848083   

income_type    paternity / maternity leave    retiree     student  unemployed  
days_employed                  3296.759962  26.403609  578.751554   26.505617

[Tentukan apa yang akan kamu gunakan: rata-rata atau median. Jelaskan mengapa.]

In [70]:
# Mari tulis fungsi yang menghitung rata-rata atau median (tergantung keputusanmu) berdasarkan parameter yang kamu identifikasi
def group_fillna(dataframe, group_col, missing_col):
    median = pd.pivot_table(df, columns=group_col, values=missing_col, aggfunc='median')

    for col in median.columns:
        df.loc[(df[group_col]==col) & (df[missing_col].isnull()), missing_col] = median[col][0]

    return df

In [71]:
# Periksa apakah fungsimu bekerja
df.isnull().sum()


children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_category           0
dtype: int64

In [72]:
# Terapkan fungsi ke income_type
df = group_fillna(dataframe=df, group_col='income_type', missing_col='days_employed')
df.head()


children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0      24.614155         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose age_category  
0    purchase of the house        40-60  
1             car purchase        20-40  
2    purchase of the house        20-40  
3  supplementary education        20-40  
4        to have a wedding        40-60

In [71]:
# Periksa apakah fungsimu bekerja


In [72]:
# Ganti nilai yang hilang



[Setelah kamu selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

In [73]:
# Periksa entri di semua kolom - pastikan kita memperbaiki semua nilai yang hilang
df.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_category        0
dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Pengkategorian data

[Untuk menjawab pertanyaan dan menguji hipotesis, kamu akan bekerja dengan data yang telah dikategorikan. Lihat pertanyaan-pertanyaan yang diajukan kepadamu dan yang harus kamu jawab. Pikirkan tentang data mana yang perlu dikategorikan untuk menjawab pertanyaan-pertanyaan ini. Di bawah ini, kamu akan menemukan templat yang bisa kamu gunakan untuk mengkategorikan data. Proses pertama mencakup data teks; yang kedua membahas data numerik yang perlu dikategorikan. Kamu dapat menggunakan kedua petunjuk yang disarankan atau tidak sama sekali - semuanya terserahmu.]

[Terlepas dari keputusanmu untuk mengatasi pengkategorian, pastikan bahwa kamu secara lugas memberikan penjelasan tentang mengapa kamu membuat keputusan tersebut. Ingat: ini adalah pekerjaanmu dan hanya kamu yang berhak membuat segala keputusan.]


In [74]:
# Tampilkan nilai data yang kamu pilih untuk pengkategorian
df[['purpose']]


purpose
0        purchase of the house
1                 car purchase
2        purchase of the house
3      supplementary education
4            to have a wedding
...                        ...
21520     housing transactions
21521        purchase of a car
21522                 property
21523        buying my own car
21524             to buy a car

[21453 rows x 1 columns]

[Mari kita periksa nilai unik]

In [75]:
# Periksa nilai unik
df['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

[Kelompok utama apakah yang dapat kamu identifikasi berdasarkan nilai uniknya?]

[Kita akan mengategorikan data kita berdasarkan topik ini.]


In [76]:
# Mari kita tulis sebuah fungsi untuk mengategorikan data berdasarkan topik umum
def category_purpose (tujuan):
    tujuan = tujuan.lower()
    if ('house' in tujuan) or ('property' in tujuan) or ('estate' in tujuan) or ('housing' in tujuan):
        category = 'house'
    elif ('educat' in tujuan) or ('university' in tujuan):
        category = 'education'
    elif ('wedding' in tujuan):
        category = 'wedding'
    elif ('car' in tujuan):
        category = 'car'
    else:
        category = 'other'

    return category

In [77]:
# Buat kolom yang memuat kategori dan hitung nilainya
df['purpose_group'] = df['purpose'].apply(category_purpose)
df

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0      24.614155         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0      24.879731         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose age_category purpose_group  
0         40620.102    purchase of the house        40-60         house  
1         17932.802             car purchase        20-40           car  
2         23341.752    purchase of the house        20-40         house  
3         42820.568  supplementary education        20-40     education  
4         25378.572        to have a wedding        40-60       wedding  
...             ...                      ...          ...           ...  
21520     35966.698     housing transactions        40-60         house  
21521     24959.969        purchase of a car        60-80           car  
21522     14347.610                 property        20-40         house  
21523     39054.888        buying my own car        20-40           car  
21524     13127.587             to buy a car        20-40           car  

[21453 rows x 14 columns]

In [78]:
df['purpose_group'].value_counts()

house        10811
car           4306
education     4013
wedding       2323
Name: purpose_group, dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Jika kamu memutuskan untuk mengategorikan data numerik, kamu juga harus membuat kategori untuk data tersebut.]

In [79]:
# Lihat semua data numerik di kolom yang kamu pilih untuk pengkategorian
df['total_income']

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21520    35966.698
21521    24959.969
21522    14347.610
21523    39054.888
21524    13127.587
Name: total_income, Length: 21453, dtype: float64

In [80]:
# Dapatkan kesimpulan statistik untuk kolomnya
df['total_income'].describe()


count     21453.000000
mean      26433.263821
std       15689.407478
min        3306.762000
25%       17219.352000
50%       23219.807500
75%       31331.009000
max      362496.645000
Name: total_income, dtype: float64

[Tentukan rentang apa yang akan kamu gunakan untuk pengelompokan dan jelaskan alasannya.]

In [81]:
# Buat fungsi yang melakukan pengkategorian menjadi kelompok numerik yang berbeda berdasarkan rentang
def income_category(income):
    if income <= 5000:
        category = '0-5000'
    elif income > 5000 and income <= 25000:
        category = '5000-25000'
    elif income > 25000 and income <= 50000:
        category = '25000-50000'
    elif income > 50000 and income <= 100000:
        category = '50000-100000'
    else:
        category = '>100000'

    return category


In [82]:
# Buat kolom yang memuat kategori

df['category_income'] = df['total_income'].apply(income_category)
df.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0      24.614155         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose age_category purpose_group category_income  
0    purchase of the house        40-60         house     25000-50000  
1             car purchase        20-40           car      5000-25000  
2    purchase of the house        20-40         house      5000-25000  
3  supplementary education        20-40     education     25000-50000  
4        to have a wedding        40-60       wedding     25000-50000

In [83]:
# Hitung setiap nilai kategori untuk melihat pendistribusiannya
df['category_income'].value_counts()

5000-25000      12824
25000-50000      7283
50000-100000     1221
>100000            99
0-5000             26
Name: category_income, dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Memeriksa hipotesis


**Apakah terdapat korelasi antara memiliki anak dengan probabilitas melakukan gagal bayar pinjaman?**

In [84]:
# Periksa data anak dan data gagal bayar pinjaman

# Ganti value debt dari integer ke string
df['debt'] = df['debt'].replace(0, 'Lancar').replace(1, 'Gagal Bayar')
child_debt = pd.pivot_table(df,
                            index='children',
                            columns='debt',
                            values ='income_type',
                            aggfunc='count',
                            margins=True).reset_index()
child_debt = child_debt[:-1]
child_debt

# Hitung persentase gagal bayar berdasarkan jumlah anak
child_debt['% Gagal Bayar'] = child_debt['Gagal Bayar'] / child_debt['All'] * 100
child_debt['% Lancar'] = child_debt['Lancar'] / child_debt['All'] * 100
child_debt


debt children  Gagal Bayar   Lancar    All  % Gagal Bayar    % Lancar
0           0       1063.0  13027.0  14090       7.544358   92.455642
1           1        445.0   4410.0   4855       9.165808   90.834192
2           2        202.0   1926.0   2128       9.492481   90.507519
3           3         27.0    303.0    330       8.181818   91.818182
4           4          4.0     37.0     41       9.756098   90.243902
5           5          NaN      9.0      9            NaN  100.000000

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatan yang kamu lakukan.]

Berdasarkan data 'children', semakin banyak jumlah anak maka probabilitas Gagal Bayar semakin besar.


<div class="alert alert-block alert-info>
Kesimpulan berdasarkan data 'children', semakin banyak anak, persentase Gagal Bayar semakin besar
            
<div/>
            

**Apakah terdapat korelasi antara status keluarga dengan probabilitas melakukan gagal bayar pinjaman?**

In [85]:
# Periksa data status keluarga dan data gagal bayar pinjaman
fam_debt = pd.pivot_table(df,
                            index='family_status',
                            columns='debt',
                            values ='income_type',
                            aggfunc='count',
                            margins=True).reset_index()
fam_debt = fam_debt[:-1]
fam_debt


# Hitung persentase gagal bayar berdasarkan status keluarga
fam_debt['% Gagal Bayar'] = fam_debt['Gagal Bayar'] / fam_debt['All'] * 100
fam_debt['% Lancar'] = fam_debt['Lancar'] / fam_debt['All'] * 100
fam_debt


debt      family_status  Gagal Bayar  Lancar    All  % Gagal Bayar   % Lancar
0     civil partnership          388    3762   4150       9.349398  90.650602
1              divorced           85    1110   1195       7.112971  92.887029
2               married          931   11408  12339       7.545182  92.454818
3             unmarried          274    2536   2810       9.750890  90.249110
4       widow / widower           63     896    959       6.569343  93.430657

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatanmu.]
Berdasarkan data 'family_status, perbedaan status keluarga tidak berpengaruh terhadap probabilitas gagal bayar.

**Apakah terdapat korelasi antara tingkat pendapatan dengan probabilitas melakukan gagal bayar pinjaman?**

In [86]:
# Periksa data tingkat pendapatan dan data gagal bayar pinjaman
income_debt = pd.pivot_table(df,
                            index='category_income',
                            columns='debt',
                            values ='income_type',
                            aggfunc='count',
                            margins=True).reset_index()
income_debt = income_debt[:-1]
income_debt


# Hitung persentase gagal bayar berdasarkan tingkat pendapatan
income_debt['% Gagal Bayar'] = income_debt['Gagal Bayar'] / income_debt['All'] * 100
income_debt['% Lancar'] = income_debt['Lancar'] / income_debt['All'] * 100
income_debt


debt category_income  Gagal Bayar  Lancar    All  % Gagal Bayar   % Lancar
0             0-5000            2      24     26       7.692308  92.307692
1        25000-50000          583    6700   7283       8.004943  91.995057
2         5000-25000         1064   11760  12824       8.296943  91.703057
3       50000-100000           86    1135   1221       7.043407  92.956593
4            >100000            6      93     99       6.060606  93.939394

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatanmu.]
Berdasarkan data 'category_income', perbedaan pendapatan tidak berpengaruh terhadap probabilitas gagal bayar.

**Bagaimana tujuan kredit memengaruhi persentase gagal bayar?**

In [87]:
# Periksa persentase tingkat gagal bayar untuk setiap tujuan kredit dan lakukan penganalisisan
purpose_debt = pd.pivot_table(df,
                            index='purpose_group',
                            columns='debt',
                            values ='income_type',
                            aggfunc='count',
                            margins=True).reset_index()
purpose_debt = purpose_debt[:-1]
purpose_debt

#persentase gagal bayar berdasarkan tujuan kredit
purpose_debt['% Gagal Bayar'] = purpose_debt['Gagal Bayar'] / purpose_debt['All'] * 100
purpose_debt['% Lancar'] = purpose_debt['Lancar'] / purpose_debt['All'] * 100
purpose_debt

debt purpose_group  Gagal Bayar  Lancar    All  % Gagal Bayar   % Lancar
0              car          403    3903   4306       9.359034  90.640966
1        education          370    3643   4013       9.220035  90.779965
2            house          782   10029  10811       7.233373  92.766627
3          wedding          186    2137   2323       8.006888  91.993112

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatan yang kamu lakukan.]
Berdasarkan data 'purpose_group', perbedaan tujuan kredit tidak berpengaruh terhadap probabilitas gagal bayar.


<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

# Kesimpulan umum 

[Tuliskan kesimpulanmu di bagian akhir ini. Pastikan kamu memasukkan semua kesimpulan penting yang telah kamu buat berkaitan dengan caramu memproses dan menganalisis data. Kesimpulan tersebut harus membahas nilai yang hilang, duplikat, dan kemungkinan alasan, serta solusi untuk data bermasalah yang harus kamu tangani.]

[Tuliskan juga kesimpulanmu mengenai pertanyaan-pertanyaan yang diajukan di sini.]


Kesimpulan Umum

1. Terdapat 12 kolom dan yang mengandung nilai yang hilang hanya terdapat di dua kolom yakni kolom 'days_employed' dan 'total income'.
2. Distribusi nilai yang hilang di kolom 'days_employed' dan 'total_income' simetris karena nilai dalam dataset dengan nilai yang telah difilter nilainya sama.
3. Nilai di dalam kolom 'days_employed' masih dalam satuan jam sehingga harus dibagi dengan 24 untuk mendapatkan nilai harinya (days).
4. Data di dalam kolom 'education' masih belum seragam sehingga harus diseragamkan dulu dengan cara diubah ke huruf kecil semua.
5. Data di kolom 'children' masih ada yang aneh yaitu angka 20 dan -1. Kemungkinan hal ini terjadi karena salah input. Maka nilai 20 diganti dengan 2 dan nilai -1 diganti dengan 1.
6. Nilai yang hilang dalam kolom 'days_employed' dan 'total_income' diubah menjadi nilai median karena ada nilai outlier.
7. Data di dalam kolom 'purpose' masih banyak mengandung nilai duplikat implisit (penulisan beda tapi maksudnya sama), sehingga harus dikategorikan dulu menjadi 4 kategori yakni: Car, Education, House, Wedding
8. Jumlah anak berpengaruh pada probabilitas gagal bayar. Semakin banyak anak membuat probabilitas gagal bayar semakin besar.
9. Status keluarga, tingkat pendapatan, dan tujuan kredit tidak berpengaruh pada probabilitas gagal bayar.

<div class="alert alert-success">
<b>Code Reviewers's comment v.2</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-block alert-warning">

# General comments #1
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Good Job Ahmad! Sebagian besar kamu telah menyelesaikan project ini dengan baik namun masih ada beberapa hal yang bisa diimprove yang saya tandai dengan comment berwarna kuning dan merah. Jika kamu telah merevisinya, projectmu akan di approve dan kamu bisa lanjut ke tahapan selanjutnya!. Beberapa revisi tersebut adalah
    
- Nilai yang aneh pada kolom days_employed
- Pembuatan kesimpulan umum
    
Keep going Ahmad, you are doing very well!
    

</div>

<div class="alert alert-block alert-success">

# General comments #2
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Good Job Ahmad! Revisi sudah sesuai ya!
    

</div>